src/util_i_variables.nss

```diff
@@ -0,0 +1,2051 @@
+    return q;
+}
+
+/// @brief Wrapper for ComplexDelete
+sqlquery _PrepareVariableDeleteAll(object oObject, string sTag, int bCampaign)
+{
+    return _PrepareComplexVariableDelete(oObject, 0, "", sTag, 0, bCampaign);
+}
+
+/// @brief Wrapper for ComplexDelete
+sqlquery _PrepareVariableDeleteByName(object oObject, string sVarName, string sTag, int bCampaign)
+{
+    return _PrepareComplexVariableDelete(oObject, 0, sVarName, sTag, 0, bCampaign);
+}
+
+/// @brief Wrapper for ComplexDelete
+sqlquery _PrepareVariableDeleteByType(object oObject, int nType, string sTag, int bCampaign)
+{
+    return _PrepareComplexVariableDelete(oObject, nType, "", sTag, 0, bCampaign);
+}
+
+/// @brief Wrapper for ComplexDelete
+sqlquery _PrepareVariableDeleteByTime(object oObject, int nTime, string sTag, int bCampaign)
+{
+    return _PrepareComplexVariableDelete(oObject, 0, "", sTag, nTime, bCampaign);
+}
+
+/// @brief Increments/Decremenst an existing variable (int/float).  If the variable
+///     does not exist, creates variables, then increments/decrements.
+sqlquery _PrepareVariableIncrement(object oObject, string sVarName, string sTag, int bCampaign)
+{
+    int bPC = GetIsPC(oObject);
+    string sTable = bPC ? VARIABLE_TABLE_PC : VARIABLE_TABLE_MODULE;
+
+    string s =  "INSERT INTO " + sTable + " " +
+                    "(type, varname, value, tag, timestamp) " +
+                "VALUES (@type, @varname, @value, @tag, strftime('%s','now')) " +
+                "ON CONFLICT (type, varname) " +
+                    "DO UPDATE SET value = value + @value, " +
+                        "timestamp = strftime('%s','now') " +
+                "RETURNING value;";
+
+    sqlquery q = bPC || !bCampaign ? _PrepareQueryObject(oObject, s) : _PrepareQueryCampaign(s);
+    SqlBindString(q, "@varname", sVarName);
+    SqlBindString(q, "@tag", sTag);
+    return q;
+}
+
+/// @brief Appends a string to an existing variable.  If the variables does not
+///     exist, creates the variable, then appends.
+sqlquery _PrepareVariableAppend(object oObject, string sVarName, string sTag, int bCampaign)
+{
+    int bPC = GetIsPC(oObject);
+    string sTable = bPC ? VARIABLE_TABLE_PC : VARIABLE_TABLE_MODULE;
+
+    string s =  "INSERT INTO " + sTable + " " +
+                    "(type, varname, value, tag, timestamp) " +
+                "VALUES (@type, @varname, @value, @tag, strftime('%s', 'now')) " +
+                "ON CONFLICT (type, varname) " +
+                    "DO UPDATE SET value = value || @value, " +
+                        "timestamp = strftime('%s', 'now') " +
+                "RETURNING value;";
+
+    sqlquery q = bPC || !bCampaign ? _PrepareQueryObject(oObject, s) : _PrepareQueryCampaign(s);
+    SqlBindString(q, "@varname", sVarName);
+    SqlBindString(q, "@tag", sTag);
+    return q;
+}
+
+/// @brief Returns a json of variable values or json objects containing
+///     variable metadata.
+json _GetVariablesByPattern(sqlquery q, int nType)
```

Comment:
Why have two separate return behaviors based on whether only one variable type was passed? If the user asks for, say, all ints with a certain varname or tag pattern, they will have to execute a separate query to go get the varname or tag.

Reply:
That's a good point.  I struggled a bit with the correct implementation of this capability.  I wanted to be able to return a simple json list of value for easy looping once data is retrieved (i.e. looping through the json array, without having to pull out values from key:value pairs), but also needed to ensure the user knows the data construct when they're pulling variables of more than one data type.  Do you think a better implementation would be to have a single behavior where it always returns the full object array, or give the user the option when returning only one type to just get the values?  Or, if two behaviors are desired, split the behaviors into separate functions?

Reply:
I think returning the full array every time is probably enough. If someone is using this function, they're likely wanting to do something complicated and won't mind the bit of extra work to access the values.

Reply:
Wilco, will modify to a single return type in the next commit.  This will also resolve the `<=0` comment.